### PR TITLE
add reference to pcidrv and pcidev

### PR DESCRIPTION
**Pull request by maxzhen**
#### Problem solved by the commit
Need to add below reference since objects can be freed our of order:
pcidev -> pcidrv
device_linux -> pcidev
#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
N/A
#### How problem was solved, alternative solutions (if any) and why they were rejected
N/A
#### Risks (if any) associated the changes in the commit
Low risk since there is no code logic change.
#### What has been tested and how, request additional testing if necessary
Build and tested with xbutil validate.
Manually tested the reference by hacking up code.
#### Documentation impact (if any)
N/A

### CODE REVIEW

File: src/runtime_src/core/pcie/linux/device_linux.h [hunk 1]
```diff
@@ -86,9 +93,16 @@ class device_linux : public shim<device_pcie>
...
     return xrt::shim_int::alloc_bo(get_device_handle(), userptr, size, xcl_bo_flags{flags}.flags);
   }
-  ////////////////////////////////////////////////////////////////
+
+protected:
+  std::shared_ptr<pci::dev>
+  get_dev() const
+  {
+    return m_pcidev;
+  }
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/device_linux.h:102]: Suggest to return `pci::dev*` and not the shared pointer
```suggestion
  pci::dev*
  get_dev() const
  {
    return m_pcidev.get();
  }
```
The shared_ptr is for sharing ownership, not of the class I see to this function needs to share ownership of the pcidev.

File: src/runtime_src/core/pcie/linux/pcidev.cpp [hunk 2]
```diff
@@ -455,21 +455,23 @@ open(const std::string& subdev, int flag) const
 }
 
 dev::
-dev(const drv& driver, const std::string& sysfs) : m_sysfs_name(sysfs), m_driver(driver)
+dev(std::shared_ptr<const drv> driver, const std::string& sysfs)
+  : m_sysfs_name(sysfs)
+  , m_driver(driver)
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/pcidev.cpp:460]:
```suggestion
dev(std::shared_ptr<const drv> driver, std::string sysfs)
  : m_sysfs_name(std::move(sysfs))
  , m_driver(std::move(driver))
```

File: src/runtime_src/core/pcie/linux/pcidev.cpp [hunk 3]
```diff
@@ -455,21 +455,23 @@ open(const std::string& subdev, int flag) const
...
 
   if(sscanf(sysfs.c_str(), "%hx:%hx:%hx.%hx", &m_domain, &m_bus, &m_dev, &m_func) < 4)
     throw std::invalid_argument(sysfs + " is not valid BDF");
 
-  m_is_mgmt = !driver.is_user();
+  m_is_mgmt = !driver->is_user();
 
   if (m_is_mgmt) {
     sysfs_get("", "instance", err, m_instance, static_cast<uint32_t>(INVALID_ID));
   } else {
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/pcidev.cpp:471]:
```suggestion
  } 
  else {
```

File: src/runtime_src/core/pcie/linux/pcidev.h [hunk 4]
```diff
@@ -185,12 +185,12 @@ class dev
 
 private:
   int
-  map_usr_bar();
+  map_usr_bar() const;
 
-  std::mutex m_lock;
-  char *m_user_bar_map = reinterpret_cast<char *>(MAP_FAILED);
+  mutable std::mutex m_lock;
+  mutable char *m_user_bar_map = reinterpret_cast<char *>(MAP_FAILED);
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/pcidev.h:191]: While mutable mutexes are obvious, m_user_bar_map is less so.  Maybe comment what it is used for and why it is made useable and perceived as constant by const member functions.

**Review by stsoe — CHANGES_REQUESTED** [review]
Looks good.  Few requests.

File: src/runtime_src/core/pcie/linux/pcidev.cpp (hunk 2 above)
Comment (maxzhen) [reply src/runtime_src/core/pcie/linux/pcidev.cpp:460]: Can't make this change. Xbmgmt or xbutil will segfault. This is the constructor of pcidev object, caller may not expect that the arguments are std::move'd? Say, caller passes in a 'driver', to construct pcidev, then use the same 'driver' to do something else...

File: src/runtime_src/core/pcie/linux/device_linux.h (hunk 1 above)
Comment (maxzhen) [reply src/runtime_src/core/pcie/linux/device_linux.h]: fixed.

File: src/runtime_src/core/pcie/linux/pcidev.cpp (hunk 3 above)
Comment (maxzhen) [reply src/runtime_src/core/pcie/linux/pcidev.cpp]: fixed.

File: src/runtime_src/core/pcie/linux/pcidev.h (hunk 4 above)
Comment (maxzhen) [reply src/runtime_src/core/pcie/linux/pcidev.h:192]: comment added.

File: src/runtime_src/core/pcie/linux/pcidev.cpp (hunk 2 above)
Comment (stsoe) [reply src/runtime_src/core/pcie/linux/pcidev.cpp:460]: Both arguments are passed by value.  `driver` you already passed by value and I changed `sysfs` to by value, so yes, you can and should move them into the data members.  By-value-and-move is btw the recommend way to initialize data member.

**Review by stsoe — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/pcie/linux/pcidev.cpp (hunk 2 above)
Comment (maxzhen) [reply src/runtime_src/core/pcie/linux/pcidev.cpp]: You're right. It is the implementation of the constructor which is not ready for the by-value-and-move change. Now it's fixed.

**Review by stsoe — APPROVED** [review]

**Pull request merged**